File: packs/security/encoders/paloalto.py
"""Palo Alto Networks traffic encoder."""

from __future__ import annotations

from eventweave.core.event import Event
from eventweave.encoders import Encoder, EncodeResult, encoder
from packs.security.encoders._helpers import format_csv


@encoder("paloalto-traffic", content_type="text/plain")
class PaloAltoTrafficEncoder(Encoder):
    """Encode events as PAN-OS traffic CSV-style logs."""

    name = "paloalto-traffic"
    content_type = "text/plain"

    _required = ["receive_time", "serial", "src", "dst", "sport", "dport", "proto", "action"]
    _columns = [
        "receive_time",
        "serial",
        "type",
        "subtype",
        "src",
        "dst",
        "natsrc",
        "natdst",
        "rule",
        "app",
        "from",
        "to",
        "sport",
        "dport",
        "proto",
        "action",
        "bytes",
        "packets",
    ]

    def encode(self, event: Event) -> EncodeResult:
        missing = [f for f in self._required if f not in event.attributes]
        if missing:
            return self._fail(f"missing required fields: {', '.join(missing)}")

        values = [event.attributes.get(col) for col in self._columns]
        return self._ok(format_csv(values))

File: runtime-go/internal/encoder/security/paloalto.go
package security

import (
	"fmt"

	"github.com/linxinhong/eventweave/runtime-go/internal/encoder"
	"github.com/linxinhong/eventweave/runtime-go/internal/event"
)

// PaloAltoTraffic encodes events as Palo Alto Networks traffic log CSV.
type PaloAltoTraffic struct{}

func init() {
	encoder.Register("paloalto-traffic", PaloAltoTraffic{})
}

// Name returns the encoder name.
func (PaloAltoTraffic) Name() string { return "paloalto-traffic" }

// ContentType returns the MIME type of the encoded output.
func (PaloAltoTraffic) ContentType() string { return "text/csv" }

// Encode formats the event as a Palo Alto traffic log CSV line.
func (PaloAltoTraffic) Encode(ev event.Event) ([]byte, error) {
	if err := require(ev, []string{"receive_time", "serial", "src", "dst", "sport", "dport", "proto", "action"}); err != nil {
		return nil, err
	}
	line := fmt.Sprintf(
		"%s,%s,%s,%s,%s,%s,%s,%s",
		asString(ev.Attributes["receive_time"]),
		asString(ev.Attributes["serial"]),
		asString(ev.Attributes["src"]),
		asString(ev.Attributes["dst"]),
		asString(ev.Attributes["sport"]),
		asString(ev.Attributes["dport"]),
		asString(ev.Attributes["proto"]),
		asString(ev.Attributes["action"]),
	)
	return []byte(line), nil
}
